File: app/qr.py
"""QR code generation for KSeF invoice verification."""

import base64
import hashlib
import io

import qrcode
import qrcode.constants

_BASE_URL = "https://qr.ksef.mf.gov.pl/invoice"


def build_verification_url(nip: str, invoice_date: str, xml_bytes: bytes) -> str:
    """Build the KSeF QR verification URL.

    Args:
        nip: Taxpayer NIP number.
        invoice_date: Invoice date in ISO format (YYYY-MM-DD).
        xml_bytes: Raw XML bytes of the invoice.

    Returns:
        URL of the form: https://qr.ksef.mf.gov.pl/invoice/{nip}/{DD-MM-YYYY}/{hash}
    """
    # SHA-256 → Base64URL (no padding)
    digest = hashlib.sha256(xml_bytes).digest()
    hash_b64url = base64.urlsafe_b64encode(digest).rstrip(b"=").decode()

    # Convert date from YYYY-MM-DD to DD-MM-YYYY
    year, month, day = invoice_date.split("-")
    date_formatted = f"{day}-{month}-{year}"

    return f"{_BASE_URL}/{nip}/{date_formatted}/{hash_b64url}"


def generate_qr_base64(url: str) -> str:
    """Generate a QR code PNG and return it as a base64 data URI.

    Args:
        url: The URL to encode in the QR code.

    Returns:
        Data URI string: "data:image/png;base64,<base64-encoded PNG>"
    """
    qr = qrcode.QRCode(
        error_correction=qrcode.constants.ERROR_CORRECT_M,
        box_size=6,
        border=2,
    )
    qr.add_data(url)
    qr.make(fit=True)

    img = qr.make_image(fill_color="black", back_color="white")
    buf = io.BytesIO()
    img.save(buf, format="PNG")
    png_bytes = buf.getvalue()

    encoded = base64.b64encode(png_bytes).decode()
    return f"data:image/png;base64,{encoded}"

File: web2/src/qr.ts
// QR code generation for KSeF invoice verification — port of app/qr.py.

import QRCode from "qrcode";

const BASE_URL = "https://qr.ksef.mf.gov.pl/invoice";

/** Base64URL (RFC 4648 §5) — like base64 but +/= replaced by -_ and padding stripped. */
function base64UrlEncode(bytes: Uint8Array): string {
  let bin = "";
  for (let i = 0; i < bytes.length; i++) bin += String.fromCharCode(bytes[i]!);
  return btoa(bin).replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

/**
 * Build the KSeF QR verification URL.
 *
 * Format: https://qr.ksef.mf.gov.pl/invoice/{nip}/{DD-MM-YYYY}/{sha256_b64url(xml)}
 */
export async function buildVerificationUrl(
  nip: string,
  invoiceDateIso: string,
  xmlBytes: Uint8Array,
): Promise<string> {
  // Copy bytes into a fresh ArrayBuffer — TS narrows Uint8Array.buffer to
  // ArrayBuffer | SharedArrayBuffer and digest only accepts the former.
  const buf = new ArrayBuffer(xmlBytes.byteLength);
  new Uint8Array(buf).set(xmlBytes);
  const digest = await crypto.subtle.digest("SHA-256", buf);
  const hash = base64UrlEncode(new Uint8Array(digest));
  const [year, month, day] = invoiceDateIso.split("-");
  const date = `${day}-${month}-${year}`;
  return `${BASE_URL}/${nip}/${date}/${hash}`;
}

/** Generate QR code as PNG data URI. */
export async function generateQrDataUrl(url: string): Promise<string> {
  return QRCode.toDataURL(url, {
    errorCorrectionLevel: "M",
    margin: 2,
    width: 270,
    color: { dark: "#000000", light: "#ffffff" },
  });
}
